### ISSUE TITLE
[FATAL] Yield coroutines while stack unwinding causes panic while panicking

### ISSUE THREAD

**Issue by zonyitoo** [zonyitoo/coio-rs#70]
Reproducible example:

```rust
extern crate coio;
extern crate env_logger;

use coio::Scheduler;

fn main() {
    env_logger::init();

    Scheduler::new()
        .run(|| {
            struct A;

            impl Drop for A {
                fn drop(&mut self) {
                    Scheduler::sched();
                }
            }

            let _a = A;

            panic!("PANICKED in coroutine");
        }).unwrap();
}
```

While the coroutine is unwinding, `Drop` of `A` will be called. Coroutine is yield in the `drop()` function. Coroutine is now suspended, which means that the currently thread is still in panicking status, but execution process is now be switched to another coroutine. If the other coroutine panic, too, then it will definitely cause panic while panicking!.

**Comment by zonyitoo** [comment]
cc @lhecker , this is the root cause.

**Comment by lhecker** [comment]
@zonyitoo Yup that's basically why we asked the Rust devs to reconsider and stop using TLS for panics in the stdlib back then when we worked on coio. Since they denied my PR and basically said that it's unlikely even for an RFC to change this behaviour I think this might be a dead end for the project. 😕 (Unless you fork the stdlib of course.)
